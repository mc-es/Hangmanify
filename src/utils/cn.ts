import type { ClassValue } from 'clsx';
import clsx from 'clsx';
import { twMerge } from 'tailwind-merge';

/**
 * ### Combines conditional Tailwind class names and resolves conflicts
 *
 * @param {...ClassValue[]} inputs - Class names, booleans, objects, or arrays
 * @returns {string} Merged and deduplicated class string
 * @example
 * cn('p-4', isDark && 'bg-black') // 'p-4 bg-black' if isDark is true, otherwise 'p-4'
 */
export const cn = (...inputs: ClassValue[]): string => twMerge(clsx(...inputs));
