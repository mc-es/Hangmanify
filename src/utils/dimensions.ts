import { Dimensions as NativeDimensions, PixelRatio } from 'react-native';

const { width: deviceWidth, height: deviceHeight } = NativeDimensions.get('window');

const guidelineBaseWidth = 375;
const guidelineBaseHeight = 812;
const BLEND_ZONE = 5;

const BREAKPOINTS = {
  '2xs': { size: 320, scale: 0.65 },
  'xs-2xs': { size: 340, scale: 0.7 },
  xs: { size: 360, scale: 0.75 },
  'xs-sm': { size: 375, scale: 0.75 },
  sm: { size: 390, scale: 0.85 },
  'sm-md': { size: 420, scale: 0.9 },
  md: { size: 480, scale: 1 },
  'md-lg': { size: 540, scale: 1.05 },
  lg: { size: 600, scale: 1.1 },
  'lg-xl': { size: 680, scale: 1.15 },
  xl: { size: 760, scale: 1.2 },
  'xl-2xl': { size: 800, scale: 1.3 },
  '2xl': { size: 840, scale: 1.4 },
} as const;

const easeIn = (t: number): number => t * t;
const easeOut = (t: number): number => t * (2 - t);

const getScale = (width: number): number => {
  const breakpoints = Object.values(BREAKPOINTS).sort((x, y) => x.size - y.size);

  if (width <= breakpoints[0].size) return breakpoints[0].scale;
  if (width >= breakpoints[breakpoints.length - 1].size)
    return breakpoints[breakpoints.length - 1].scale;

  for (let i = 0; i < breakpoints.length - 1; i++) {
    const current = breakpoints[i];
    const next = breakpoints[i + 1];

    if (width >= current.size && width < next.size) {
      const scaleRange = next.scale - current.scale;
      const widthRange = next.size - current.size;
      const delta = width - current.size;

      let finalProgress;

      if (delta < BLEND_ZONE) {
        const t = delta / BLEND_ZONE;
        finalProgress = easeIn(t) * (BLEND_ZONE / widthRange);
      } else if (delta > widthRange - BLEND_ZONE) {
        const t = (delta - (widthRange - BLEND_ZONE)) / BLEND_ZONE;
        finalProgress =
          (widthRange - BLEND_ZONE) / widthRange + easeOut(t) * (BLEND_ZONE / widthRange);
      } else finalProgress = delta / widthRange;

      return current.scale + scaleRange * finalProgress;
    }
  }

  return 1;
};

const scale = getScale(deviceWidth);

// horizontal & vertical scale
const hs = (size: number): number => (deviceWidth / guidelineBaseWidth) * size * scale;
const vs = (size: number): number => (deviceHeight / guidelineBaseHeight) * size;

// width & height percent
const wp = (percent: number): number => (deviceWidth * percent) / 100;
const hp = (percent: number): number => (deviceHeight * percent) / 100;

// moderate scale
const ms = (size: number, factor = 0.5): number => size + (hs(size) - size) * factor;

// font scale
const fs = (size: number): number => {
  const fontScale = PixelRatio.getFontScale();
  return Math.round(PixelRatio.roundToNearestPixel(hs(size) * fontScale));
};

export const Dimensions = { hs, vs, wp, hp, ms, fs };
