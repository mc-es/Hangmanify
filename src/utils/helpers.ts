import type { TranslateOptions } from 'i18n-js';

function getNestedValue<T>(obj: T, key: string): unknown {
  return key.split('.').reduce((acc: unknown, curr: string): unknown => {
    if (acc !== null && typeof acc === 'object' && curr in acc)
      return (acc as Record<string, unknown>)[curr];

    return null;
  }, obj as unknown);
}
function interpolate(template: string, options: TranslateOptions): string {
  return template.replace(/{{\s*(\w+)\s*}}/g, (_match, key) =>
    key in options ? String(options[key]) : ''
  );
}

export const helpers = { getNestedValue, interpolate };
