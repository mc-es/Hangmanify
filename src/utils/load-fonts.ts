import * as Font from 'expo-font';

import { APP_FONTS } from 'assets/fonts';

import { FONTS } from 'src/constants/styles';

export const LoadFonts = async (): Promise<void> => {
  try {
    await Font.loadAsync({
      [FONTS.Nunito.Bold]: APP_FONTS.nunito.bold,
      [FONTS.Nunito.Light]: APP_FONTS.nunito.light,
      [FONTS.Nunito.Medium]: APP_FONTS.nunito.medium,
      [FONTS.Nunito.Regular]: APP_FONTS.nunito.regular,
      [FONTS.Nunito.SemiBold]: APP_FONTS.nunito.semibold,
      [FONTS.Poppins.Bold]: APP_FONTS.poppins.bold,
      [FONTS.Poppins.Light]: APP_FONTS.poppins.light,
      [FONTS.Poppins.Medium]: APP_FONTS.poppins.medium,
      [FONTS.Poppins.Regular]: APP_FONTS.poppins.regular,
      [FONTS.Poppins.SemiBold]: APP_FONTS.poppins.semibold,
    });
  } catch (error) {
    console.error('Error loading fonts', error);
  }
};
