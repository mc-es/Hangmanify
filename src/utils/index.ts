import { cn } from './cn';
import { dimensions } from './dimensions';
import { helpers } from './helpers';

export { cn, dimensions, helpers };
