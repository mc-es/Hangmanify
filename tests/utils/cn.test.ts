import { cn } from 'src/utils/cn';

const edgeCases = [
  {
    name: 'merges simple class names',
    code: () => cn('p-4', 'text-red-500'),
    expected: 'p-4 text-red-500',
  },
  {
    name: 'ignores falsey values',
    code: () => cn('p-4', null, undefined),
    expected: 'p-4',
  },
  {
    name: 'adds class conditionally (true)',
    code: () => {
      const isDark = true;
      return cn('p-4', isDark && 'bg-black');
    },
    expected: 'p-4 bg-black',
  },
  {
    name: 'ignores conditional (false)',
    code: () => {
      const isLight = false;
      return cn('p-4', isLight && 'bg-white');
    },
    expected: 'p-4',
  },
  {
    name: 'resolves tailwind conflicts',
    code: () => cn('p-4', 'p-2'),
    expected: 'p-2',
  },
  {
    name: 'supports object syntax',
    code: () => cn({ 'bg-red-500': true, 'bg-blue-500': false }),
    expected: 'bg-red-500',
  },
  {
    name: 'returns empty string with only falsey values',
    code: () => cn(false, null, undefined),
    expected: '',
  },
];

describe('cn utility function', () => {
  edgeCases.forEach(({ name, code, expected }) => {
    it(name, () => {
      const result = code();
      expect(result).toBe(expected);
      expect(result).toMatchSnapshot();
    });
  });
});
