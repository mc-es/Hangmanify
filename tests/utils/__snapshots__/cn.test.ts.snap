// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cn utility function adds class conditionally (true) 1`] = `"p-4 bg-black"`;

exports[`cn utility function ignores conditional (false) 1`] = `"p-4"`;

exports[`cn utility function ignores falsey values 1`] = `"p-4"`;

exports[`cn utility function merges simple class names 1`] = `"p-4 text-red-500"`;

exports[`cn utility function resolves tailwind conflicts 1`] = `"p-2"`;

exports[`cn utility function returns empty string with only falsey values 1`] = `""`;

exports[`cn utility function supports object syntax 1`] = `"bg-red-500"`;
